### ISSUE TITLE
Problem: TestPushPullChanneler randomly fails (crash)

### ISSUE THREAD

**Issue by cyrilleverrier** [zeromq/goczmq#201]
Steps to reproduce:
`go test -v -run TestPushPullChanneler`

This test randomly crashes with the following callstack:

➜  goczmq git:(master) ✗ go test -v -run TestPushPullChanneler
```
=== RUN   TestPushPullChanneler
goczmq.test: src/zsys.c:179: zsys_init: Assertion `!s_process_ctx' failed.
SIGABRT: abort
PC=0x7f1904a71418 m=4
signal arrived during cgo execution

goroutine 23 [syscall, locked to thread]:
runtime.cgocall(0x582480, 0xc8200cad18, 0xc800000000)
	/usr/lib/go-1.6/src/runtime/cgocall.go:123 +0x11b fp=0xc8200cacc0 sp=0xc8200cac90
github.com/goczmq._Cfunc_zsock_new_checked(0x8, 0x7f18f80008c0, 0xc3, 0x0)
	github.com/goczmq/_test/_obj_test/_cgo_gotypes.go:924 +0x42 fp=0xc8200cad18 sp=0xc8200cacc0
github.com/goczmq.NewSock(0x8, 0x0)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/sock.go:80 +0x1f0 fp=0xc8200cada8 sp=0xc8200cad18
github.com/goczmq.NewPush(0xc820078a50, 0x21, 0x0, 0x0, 0x0)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/sock.go:195 +0x2c fp=0xc8200cade8 sp=0xc8200cada8
github.com/goczmq.(*Channeler).channeler(0xc82006c380, 0xc820070420, 0xc820070480)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/channeler.go:128 +0x4d fp=0xc8200caf98 sp=0xc8200cade8
runtime.goexit()
	/usr/lib/go-1.6/src/runtime/asm_amd64.s:1998 +0x1 fp=0xc8200cafa0 sp=0xc8200caf98
created by github.com/goczmq.newChanneler
	/home/cyrille/code/go/workspace/src/github.com/goczmq/channeler.go:185 +0x3f2

goroutine 1 [chan receive]:
testing.RunTests(0x670590, 0x916c20, 0x4c, 0x4c, 0xc820072801)
	/usr/lib/go-1.6/src/testing/testing.go:583 +0x8d2
testing.(*M).Run(0xc82004def8, 0xf)
	/usr/lib/go-1.6/src/testing/testing.go:515 +0x81
main.main()
	github.com/goczmq/_test/_testmain.go:236 +0x117

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/lib/go-1.6/src/runtime/asm_amd64.s:1998 +0x1

goroutine 20 [chan send]:
github.com/goczmq.TestPushPullChanneler(0xc8200a0360)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/channeler_test.go:17 +0x1c2
testing.tRunner(0xc8200a0360, 0x916ce0)
	/usr/lib/go-1.6/src/testing/testing.go:473 +0x98
created by testing.RunTests
	/usr/lib/go-1.6/src/testing/testing.go:582 +0x892

goroutine 21 [runnable, locked to thread]:
github.com/goczmq._Cfunc_zsock_new_checked(0x8, 0x7f18f40008c0, 0xc3, 0x7f18f4000910)
	github.com/goczmq/_test/_obj_test/_cgo_gotypes.go:924 +0x42
github.com/goczmq.NewSock(0x8, 0x0)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/sock.go:80 +0x1f0
github.com/goczmq.NewPush(0xc8200789f0, 0x21, 0x4922be, 0x0, 0x0)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/sock.go:195 +0x2c
github.com/goczmq.(*Channeler).channeler(0xc82006c310, 0xc820070300, 0xc820070360)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/channeler.go:128 +0x4d
created by github.com/goczmq.newChanneler
	/home/cyrille/code/go/workspace/src/github.com/goczmq/channeler.go:185 +0x3f2

goroutine 22 [runnable, locked to thread]:
github.com/goczmq._Cfunc_free(0x7f18fc000a20)
	github.com/goczmq/_test/_obj_test/_cgo_gotypes.go:260 +0x36
github.com/goczmq.NewSock(0x0, 0xc820014280)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/sock.go:81 +0x21a
github.com/goczmq.NewPair(0xc82000a180, 0x29, 0xc82003de40, 0x0, 0x0)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/sock.go:230 +0x2c
github.com/goczmq.(*Channeler).actor(0xc82006c310, 0xc8200703c0)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/channeler.go:35 +0x13c
created by github.com/goczmq.newChanneler
	/home/cyrille/code/go/workspace/src/github.com/goczmq/channeler.go:186 +0x41e

goroutine 24 [runnable, locked to thread]:
github.com/goczmq._Cfunc_zsock_new_checked(0x0, 0x16e5330, 0xe6, 0x16e4c20)
	github.com/goczmq/_test/_obj_test/_cgo_gotypes.go:924 +0x42
github.com/goczmq.NewSock(0x0, 0x0)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/sock.go:80 +0x1f0
github.com/goczmq.NewPair(0xc820078a80, 0x29, 0xc820039e40, 0x0, 0x0)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/sock.go:230 +0x2c
github.com/goczmq.(*Channeler).actor(0xc82006c380, 0xc8200704e0)
	/home/cyrille/code/go/workspace/src/github.com/goczmq/channeler.go:35 +0x13c
created by github.com/goczmq.newChanneler
	/home/cyrille/code/go/workspace/src/github.com/goczmq/channeler.go:186 +0x41e

rax    0x0
rbx    0x7f19054de000
rcx    0x7f1904a71418
rdx    0x6
rdi    0x7f72
rsi    0x7f77
rbp    0x7f19050adfb5
rsp    0x7f1902ac4bf8
r8     0x7f18f8000990
r9     0xfefefefeff092d00
r10    0x8
r11    0x206
r12    0xb3
r13    0x7f19050af8d8
r14    0x3
r15    0x8
rip    0x7f1904a71418
rflags 0x206
cs     0x33
fs     0x0
gs     0x0
exit status 2
FAIL	github.com/goczmq	0.006s
```

**Comment by cyrilleverrier** [comment]
### Explanation:

When creating a channel, several Go `Sock` are created in 2 different threads (go routine) at the same time:
```
func newChanneler(sockType int, endpoints, subscribe string) *Channeler {
...
	go c.channeler(commandChan, sendChan)
	go c.actor(recvChan)
...
}
```

The creation of new Go `Sock` leads to the call of `zsys_socket` in their respective go routines/C threads.

The implementation of  `zsys_socket` is as follows:
```
void * zsys_socket (int type, const char *filename, size_t line_nbr)
{
    //  First time initialization; if the application is mixing
    //  its own threading calls with zsock, this may fail if two
    //  threads try to create sockets at the same time. In such
    //  apps, they MUST create a socket in the main program before
    //  starting any threads. If the app uses zactor for its threads
    //  then we can guarantee this to always be safe.
    zsys_init ();
``

**Comment by cyrilleverrier** [comment]
A simple patch would be to call `zsys_init` earlier in the main thread, as explained by:
```

//  --------------------------------------------------------------------------
//  Initialize CZMQ zsys layer; this happens automatically when you create
//  a socket or an actor; however this call lets you force initialization
//  earlier, so e.g. logging is properly set-up before you start working.
//  Not threadsafe, so call only from main thread. Safe to call multiple
//  times. Returns global CZMQ context.
void * zsys_init (void)
```

**Comment by cyrilleverrier** [comment]
See tentative fix: PR #202

Please review it carefully as I don't have much experience with zmq/czmq/goczmq

**Comment by taotetek** [comment]
@cyrilleverrier I believe you are correct - merged, and ran tests in a loop for a bit, and looking good.  Thanks!
